src/compiler/checker.ts

```diff
@@ -32791,8 +32810,38 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function checkNonNullAssertion(node: NonNullExpression) {
-        return node.flags & NodeFlags.OptionalChain ? checkNonNullChain(node as NonNullChain) :
-            getNonNullableType(checkExpression(node.expression));
+        const exprType = checkExpression(node.expression);
+        const resultType = node.flags & NodeFlags.OptionalChain ? checkNonNullChain(node as NonNullChain) : getNonNullableType(exprType);
+        if (!isErrorType(resultType)) {
+            addLazyDiagnostic(() => {

```

Comment:
Because we _should_ be using it for _any_ diagnostic work (suggestions included), so we skip and defer the work when we do stuff like find-all-refs in the language service. We just don't because it'd be effort to audit every call of `error`, for an unknown services performance benefit.